tests/python/pants_test/integration/pep_561_integration_test.py

```diff
@@ -0,0 +1,83 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from glob import glob
+from textwrap import dedent
+
+from pants.base.build_environment import get_buildroot
+from pants.testutil.pants_integration_test import PantsResult, run_pants, setup_tmpdir
+
+
+def typecheck_file(filename: str) -> PantsResult:
+    return run_pants(
+        [
+            "--backend-packages=["
+            "'pants.backend.python',"
+            "'pants.backend.python.typecheck.mypy'"
+            "]",
+            "typecheck",
+            filename,
+        ]
+    )
+
+
+def test_typecheck_works() -> None:
+    project = {
+        "proj1/BUILD": "python_library()",
+        "proj1/ok.py": dedent(
+            """\
+            def dummy() -> int:
+                return 42
+            """
+        ),
+        "proj1/err.py": dedent(
+            """\
+            def dummy() -> int:
+                return "not an int"
+            """
+        ),
+    }
+
+    with setup_tmpdir(project) as tmpdir:
+        ok_result = typecheck_file(f"{tmpdir}/proj1/ok.py")
+        err_result = typecheck_file(f"{tmpdir}/proj1/err.py")
+
+    ok_result.assert_success()
+    err_result.assert_failure()
+
+
+def test_type_stubs() -> None:
+    wheel = glob(f"{get_buildroot()}/pantsbuild.pants.testutil-*.whl")[0]
+    project = {
+        "proj1/BUILD": dedent(
+            f"""\
+            python_requirement_library(
+                name="pants",
+                requirements=["Pants @ file://{wheel}"],
```

Comment:
@jsirois I came up with this instead of your suggestion with subprocess and stuff... didn't really get what I should do with that tbh :P

Reply:
How the heck is that glob up on line 50 working? As far as I can tell this test does nothing to set up / ensure that file's existence.

Reply:
It is in the `BUILD` file, with a `runtime_package_dependencies` on the testutil package.

Reply:
Aha - OK. Still clear as mud though. How is it that the runtime package dependency is found in the build root?! That can't be right leaking artifacts up there. As I read it the artifact should be in the test sandbox root, which is a `$TMPDIR/process-execution*/` dir.

Reply:
Oh, OK, I wouldn't know about that. From what I gathered, the field is a test construct to inject packages into tests.

I guess it's a plugin rule for pytest: https://github.com/pantsbuild/pants/blob/8b741e2087bd59233b4de2d5c11524a18000799a/src/python/pants/backend/python/goals/pytest_runner.py#L360

https://github.com/pantsbuild/pants/blob/8b741e2087bd59233b4de2d5c11524a18000799a/src/python/pants/backend/python/goals/pytest_runner_integration_test.py#L461-L463

Reply:
Sure - I get that. So I'm asking why is the test - apparently - running in the build root?. It must be for your glob to be working and "... the field is a test construct to inject packages into tests ...". That should be disturbing.

Reply:
Well, the build root seems to point to the sandbox:
```
WHEEL: /private/var/folders/8j/c8jf_msj009947wyw82xvdkw0000gn/T/process-executionEcPlB6/pantsbuild.pants.testutil-2.8.0.dev0-py37.py38.py39-none-any.whl 
```

About injecting packages being disturbing, I suppose it depends on what you inject, and for what purpose. It could be abused, for sure :P

Reply:
To be very clear, I am not disturbed about the package injection. I'm disturbed by the glob working - i.e.: the apparanet location of the injection, which is supposed to be `get_buildroot` relative according to the test code, but the build root should be up where pants.toml and the pants script live. If so, that's a leak of random test files into the repo root and that's not good at all. If not, then somehow `get_buildroot` is pointing into your test sandbox chroot, which is good - no leaks - but what's not good is nowhere in this test file do you configure the test sandbox chroot to be the build root. At least as far as I can tell. If you agree with everything I've written here, then the test is working, but by luck, since neither of us understands how the test sandbox root is magically becoming the build root for the duration of this test.

Reply:
OK, It's because the testutil wheel depends on pants_integration_test - which makes sense - we want all the test infra code shipped in the wheel:
https://github.com/pantsbuild/pants/blob/89361215073b86a36287683ebf6700e94f1c13c5/src/python/pants/testutil/BUILD#L4-L10

And pants_integration_test depends on the BUILD_ROOT file at the root of the repo:
https://github.com/pantsbuild/pants/blob/89361215073b86a36287683ebf6700e94f1c13c5/src/python/pants/testutil/BUILD#L37-L44

So that BUILD_ROOT file gets added to the test sandbox chroot.

It seems broken that the `run_pants` and `rule_runner` code don't explicitly emit a `BUILD_ROOT` file in PWD as part of their setup and instead sneak in the build root this way. The sign this is broken is that the testutil wheel should have a BUILD_ROOT file in it but does not for unclear reasons (to me). So when someone out in the world goes to use this wheel to write tests for their plugin that lives outside the Pants repo - the test will _not_ have a BUILD_ROOT file in their test sandbox and will leak out to the real build root. So this of course is a pre-existing bug that's not your problem, but your test should also probably not rely on this non-local test setup. How about switching the glob from being rooted by `get_buildroot()` to being rooted by `os.getcwd()`..

Reply:
Here's the pre-existing run_pants / RuleRunner / pantsbuild.pants.testutil dist issue: https://github.com/pantsbuild/pants/issues/12760

Reply:
Haha, ok. Thanks for those extra clarifications. Indeed, I don't fully understand everything going on, so can't always tell when things work by luck or design. The concept with `BUILD_ROOT` is only something I've seen in passing, without digging into further, so are completely unaware of what it does or how it does it.

I'll replace `get_buildroot()` with `cwd` 👍

Reply:
I'm happy you figured all that out, @jsirois, as it would have taken me considerably longer.. :D

Reply:
What I'm still a bit mystified by is how you stumbled on `get_buildroot()` at all for this. That seems like the last thing that would possibly work and that's why it set off alarm bells for me.

Reply:
In my head, buildroot == sandbox dir. Wonder if I saw someone using buildroot elsewhere, perhaps.. I'll check, see if I can find a reference such as that.

To me, buildroot feels like an ephemeral dir used during build/test, not the project root, so felt natural to go with that :P (I see now, that I was wrong.. but happy we flushed out a bug in the process..)

Reply:
Not sure I looked at it, but this seems closely related.. ?

https://github.com/pantsbuild/pants/blob/89361215073b86a36287683ebf6700e94f1c13c5/tests/python/pants_test/integration/goal_rule_integration_test.py#L53-L54

Reply:
There's quite a lot of `get_buildroot()` in test files in the source tree.. But they may actually be used correctly there, so a mis-adaptation to the integration scenario on my part, more likely.